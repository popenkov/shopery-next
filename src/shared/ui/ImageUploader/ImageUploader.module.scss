@import '@/shared/styles/mixins/flexbox';
@import '@/shared/styles/mixins/box';
@import '@/shared/styles/mixins/text-styles';

.ImageUploader {
  @include flex-column-center;

  gap: 20px;
}

.ImageUploaderPreviewWrapper {
  @include box(224px);

  position: relative;
  overflow: hidden;
  border-radius: 50%;
}

.ImageUploaderPreview {
  background-color: var(--monochrome-100);
}

.ImageUploaderInput {
  position: absolute;
  z-index: -1;
  width: 0.1px;
  height: 0.1px;
  overflow: hidden;
  opacity: 0;
}

.ImageUploaderLabel {
  @include body-s(600);

  min-height: 45px;
  padding: 14px 32px;
  color: var(--primary);
  text-decoration: none;
  border: none;
  border-radius: 43px;
  box-shadow: inset 0 0 0 1px var(--primary);
  cursor: pointer;
  transition: background-color var(--transition);

  @media (hover: hover) {
    &:hover {
      color: var(--hard-primary);
      box-shadow: inset 0 0 0 1px var(--hard-primary);
    }
  }
}
