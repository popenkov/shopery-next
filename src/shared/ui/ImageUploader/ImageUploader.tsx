'use client';

import { FC, forwardRef, useState } from 'react';
import cn from 'classnames';
import Image from 'next/image';
import { ChangeHandler } from 'react-hook-form';

import cls from './ImageUploader.module.scss';

type Props = {
  onChange: ChangeHandler;
  imageSrc: string;
  className?: string;
  accept?: string;
  maxSize?: number;
};

export const ImageUploader: FC<Props> = forwardRef<HTMLInputElement, Props>(
  ({ imageSrc, onChange, className, accept = 'image/*', maxSize = 1024 * 1024, ...rest }, ref) => {
    const [preview, setPreview] = useState<string>(imageSrc);
    const [error, setError] = useState<string | null>(null);

    const handleFileChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      const file = event.target.files?.[0];

      if (!file) return;

      if (file.size > maxSize) {
        setError(`File size exceeds ${maxSize / 1024}KB`);
        return;
      }

      onChange(event);

      const reader = new FileReader();
      reader.readAsDataURL(file);
      reader.onloadend = () => {
        setPreview(reader.result as string);
      };
    };

    return (
      <div className={cn(cls.ImageUploader, className)}>
        <div className={cls.ImageUploaderPreviewWrapper}>
          <Image
            className={cls.ImageUploaderPreview}
            fill
            src={preview}
            alt="image preview"
            data-testid="ImageUploader.preview"
          />
        </div>

        <input
          ref={ref}
          type="file"
          accept={accept}
          onChange={handleFileChange}
          className={cls.ImageUploaderInput}
          id="file"
          data-testid="ImageUploader.input"
          {...rest}
        />
        <label htmlFor="file" className={cls.ImageUploaderLabel}>
          Choose Image
        </label>
        {error && <span className={cls.ImageUploaderError}>{error}</span>}
      </div>
    );
  },
);

ImageUploader.displayName = 'ImageUploader';
