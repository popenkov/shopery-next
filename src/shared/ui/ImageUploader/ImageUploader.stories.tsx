import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';

import { AppRouteNames } from '@/shared/model';
import { ImageUploader } from './ImageUploader';

const meta = {
  title: 'shared/ImageUploader',
  component: ImageUploader,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  args: { imageSrc: 'image path' },
  argTypes: {},
} satisfies Meta<typeof ImageUploader>;

export default meta;
type TStory = StoryObj<typeof meta>;

export const ImageUploaderComponent: TStory = {
  decorators: [
    (Story) => (
      <div>
        <Story />
      </div>
    ),
  ],
  args: {
    imageSrc: '/images/product-red-capsicum.jpg',
    onChange: async (e) => console.log('onChange', e.target.files[0]),
  },
};
