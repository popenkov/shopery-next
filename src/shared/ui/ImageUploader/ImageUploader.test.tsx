import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';

import { ImageUploader } from './ImageUploader';

// eslint-disable-next-line
jest.mock('next/image', () => ({ src, alt }: { src: string; alt: string }) => (
  <img src={src} alt={alt} data-testid="ImageUploader.preview" />
));

describe('ImageUploader component', () => {
  it('renders correctly', () => {
    const imageSrc = 'https://test.com/image.jpg';
    const onChange = jest.fn();
    const { container } = render(<ImageUploader imageSrc={imageSrc} onChange={onChange} />);
    expect(container).toMatchSnapshot();
  });

  it('renders image preview correctly', () => {
    const imageSrc = 'https://test.com/image.jpg';
    const onChange = jest.fn();
    const { getByTestId } = render(<ImageUploader imageSrc={imageSrc} onChange={onChange} />);
    const imagePreview = getByTestId('ImageUploader.preview');
    expect(imagePreview).toBeInTheDocument();
  });

  it('calls onChange handler when file is selected', () => {
    const imageSrc = 'https://test.com/image.jpg';
    const onChange = jest.fn();
    const { getByTestId } = render(<ImageUploader imageSrc={imageSrc} onChange={onChange} />);
    const fileInput = getByTestId('ImageUploader.input');
    const file = new File(['test'], 'test.jpg', {
      type: 'image/jpeg',
    });
    fireEvent.change(fileInput, { target: { files: [file] } });
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  it('displays error message when file size exceeds max size', () => {
    const imageSrc = 'https://test.com/image.jpg';
    const onChange = jest.fn();
    const maxSize = 1024;
    const { getByText, getByTestId } = render(
      <ImageUploader imageSrc={imageSrc} onChange={onChange} maxSize={maxSize} />,
    );
    const fileInput = getByTestId('ImageUploader.input');
    const file = new File(['test'], 'test.jpg', {
      type: 'image/jpeg',
    });
    Object.defineProperty(file, 'size', { value: maxSize * 2 });
    fireEvent.change(fileInput, { target: { files: [file] } });
    const errorMessage = getByText(`File size exceeds ${maxSize / 1024}KB`);
    expect(errorMessage).toBeInTheDocument();
  });

  it('updates preview when file is selected', () => {
    const imageSrc = 'https://test.com/image.jpg';
    const onChange = jest.fn();
    const { getByTestId } = render(<ImageUploader imageSrc={imageSrc} onChange={onChange} />);
    const fileInput = getByTestId('ImageUploader.input');
    const file = new File(['test'], 'test.jpg', {
      type: 'image/jpeg',
    });
    fireEvent.change(fileInput, { target: { files: [file] } });
    const newPreview = 'data:image/jpeg;base64,test';
    waitFor(() => {
      const updatedPreview = getByTestId('ImageUploader.preview');
      expect(updatedPreview).toHaveAttribute('src', newPreview);
    });
  });
});
