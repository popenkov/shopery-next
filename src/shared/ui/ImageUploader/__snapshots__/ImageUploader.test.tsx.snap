// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageUploader component renders correctly 1`] = `
<div>
  <div
    class="ImageUploader"
  >
    <div
      class="ImageUploaderPreviewWrapper"
    >
      <img
        alt="image preview"
        data-testid="image-uploader-preview"
        src="https://test.com/image.jpg"
      />
    </div>
    <input
      accept="image/*"
      class="ImageUploaderInput"
      data-testid="image-uploader-input"
      id="file"
      type="file"
    />
    <label
      class="ImageUploaderLabel"
      for="file"
    >
      Choose Image
    </label>
  </div>
</div>
`;
